### ISSUE TITLE
Strangely increased producer latency due to socket.blocking.max.ms

### ISSUE THREAD

**Issue by tomciaaa** [edenhill/librdkafka#629]
HI Magnus,

Experimenting with librdkafka I noticed some peculiar behaviour: socket.blocking.max.ms value seems to introduce some blocking in other threads. More specifically, running with a batch size of 1 and waiting for ack on every single message (see [rdkafka_example_diff.txt](https://github.com/edenhill/librdkafka/files/235102/rdkafka_example_diff.txt)), I seem to run into issues with my message not being forwarded to the broker for at most socket.blocking.max.ms miliseconds.

For example, running with a 1000ms timeout, I have to poll(1), more than 800 times before I receive an ack, and the round-trip times for the Produce messages are nowhere near that: (see
[rdkafka_1s_socket_timeout_log.txt](https://github.com/edenhill/librdkafka/files/235112/rdkafka_1s_socket_timeout_log.txt) for the full log)

`for i in 1 2 3 4 5 6; do echo $i; done | ./examples/rdkafka_example_cpp -P -b localhost -t TestTopic -p 0 -X 'batch.num.messages=1' -X 'socket.blocking.max.ms=1000' -d protocol 2>&1 | uniq -c`
      1 % Produced message (1 bytes)
    880 % Waiting for ACK 
      1 LOG-7-SEND: grvldatadev1.cloudapp.net:9092/0: Sent ProduceRequest (v1, 81 bytes, CorrId 5)
      2 % Waiting for ACK 
      1 LOG-7-RECV: grvldatadev1.cloudapp.net:9092/0: Received ProduceResponse (v1, 37 bytes, CorrId 5, rtt 2.81ms)
      1 Message delivery for (1 bytes): Success

Also please note that this is unfortunately not 100% deterministic: I needed to run the test script maybe a few times before I hit the trouble scenario.

Do you know if this is the expected behaviour of socket.blocking.max.ms? Currently I am working around this by setting the value to 5 and taking the CPU penalty, as the default 100 milisecond sleeps are getting in the way of getting decent performance with a synchronous producer.

Thanks,
Tomas

**Comment by edenhill** [comment]
Hey,

I cant reproduce this off master on Linux with a localhost broker:

```
# $ for i in 1 2 3 4 5 6; do echo $i; done | examples/rdkafka_example -b 0:19172 -t test -p 0 -X 'batch.num.messages=1' -X 'socket.blocking.max.ms=1000' -d protocol -P
....

1461621756.904 RDKAFKA-7-SEND: rdkafka#producer-1: localhost:19172/19172: Sent ProduceRequest (v1, 76 bytes, CorrId 2)
1461621756.904 RDKAFKA-7-SEND: rdkafka#producer-1: localhost:19172/19172: Sent ProduceRequest (v1, 76 bytes, CorrId 3)
1461621756.904 RDKAFKA-7-SEND: rdkafka#producer-1: localhost:19172/19172: Sent ProduceRequest (v1, 76 bytes, CorrId 4)
1461621756.904 RDKAFKA-7-SEND: rdkafka#producer-1: localhost:19172/19172: Sent ProduceRequest (v1, 76 bytes, CorrId 5)
1461621756.904 RDKAFKA-7-SEND: rdkafka#producer-1: localhost:19172/19172: Sent ProduceRequest (v1, 76 bytes, CorrId 6)
1461621756.904 RDKAFKA-7-SEND: rdkafka#producer-1: localhost:19172/19172: Sent ProduceRequest (v1, 76 bytes, CorrId 7)
1461621756.905 RDKAFKA-7-RECV: rdkafka#producer-1: localhost:19172/19172: Received ProduceResponse (v1, 32 bytes, CorrId 2, rtt 1.15ms)
% Message delivered (1 bytes): 1
1461621756.905 RDKAFKA-7-RECV: rdkafka#producer-1: localhost:19172/19172: Received ProduceResponse (v1, 32 bytes, CorrId 3, rtt 1.34ms)
% Message delivered (1 bytes): 2
1461621756.905 RDKAFKA-7-RECV: rdkafka#producer-1: localhost:19172/19172: Received ProduceResponse (v1, 32 bytes, CorrId 4, rtt 1.48ms)
% Message delivered (1 bytes): 3
1461621756.906 RDKAFKA-7-RECV: rdkafka#producer-1: localhost:19172/19172: Received ProduceResponse (v1, 32 bytes, CorrId 5, rtt 1.70ms)
% Message delivered (1 bytes): 4
1461621756.906 RDKAFKA-7-RECV: rdkafka#producer-1: localhost:19172/19172: Received ProduceResponse (v1, 32 bytes, CorrId 6, rtt 2.21ms)
% Message delivered (1 bytes): 5
1461621756.906 RDKAFKA-7-RECV: rdkafka#producer-1: localhost:19172/19172: Received ProduceResponse (v1, 32 bytes, CorrId 7, rtt 2.43ms)
% Message delivered (1 bytes): 6
```

Full time for first message sent to last delivered is 2 ms

**Comment by tomciaaa** [comment]
Hi Magnus,

You seem to not have applied the patch to rdkafka_example.cpp I have attached in the previous post: [rdkafka_example_diff.txt](https://github.com/edenhill/librdkafka/files/235102/rdkafka_example_diff.txt) as well as running the C version of the example. Without the patch the producer does not block until receiving the ACK for the message and that is the scenario that produces the issue.

Also I am running this example with the current master on Ubuntu and the changes to rdkakfa_example.cpp are the only changes I have made to the code.
